### PR TITLE
Refactor persistence layer, introduce new containers

### PR DESCRIPTION
**Pull request by wxing1292**
* Introduce WorkflowMutation, as container for workflow execution updates. 
* Introduce WorkflowSnapshot, as container for workflow execution snapshots.

### CODE REVIEW

File: common/persistence/dataInterfaces.go [hunk 1]
```diff
@@ -767,60 +767,29 @@ type (
...
-
-		// replication/ transfer / timer task
-		InsertReplicationTasks []Task
-		InsertTransferTasks    []Task
-		InsertTimerTasks       []Task
+		// workflow to be resetted
+		ResetWorkflowSnapshot WorkflowSnapshot
+
+		// current workflow
+		CurrentWorkflowMutation *WorkflowMutation
```
Comment (yux0) [inline common/persistence/dataInterfaces.go]: Both WorkflowMutation and WorkflowSnapshot have ExecutionInfo/ReplicationState/VersionHistories. Which one should them be stored?

File: common/persistence/versionHistory.go [hunk 2]
```diff
@@ -458,6 +458,11 @@ func (h *VersionHistories) AddVersionHistory(
 	if newLastItem.version > currentLastItem.version {
 		currentBranchChanged = true
 		h.currentBranchIndex = newVersionHistoryIndex
+	} else if newLastItem.version == currentLastItem.version {
+		if newLastItem.eventID > currentLastItem.eventID {
+			currentBranchChanged = true
```
Comment (yux0) [inline common/persistence/versionHistory.go]: why in this case the current branch changed?

File: common/persistence/dataInterfaces.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/dataInterfaces.go]: CurrentWorkflowMutation is for the current workflow to be turned into zombie or terminated
ResetWorkflowSnapshot is the workflow which is to be resetted

File: common/persistence/versionHistory.go (hunk 2 above)
Comment (wxing1292) [reply common/persistence/versionHistory.go]: it has longer history

File: common/persistence/dataInterfaces.go [hunk 3]
```diff
@@ -767,60 +767,29 @@ type (
...
-		InsertRequestCancelInfos  []*RequestCancelInfo
-		InsertSignalInfos         []*SignalInfo
-		InsertSignalRequestedIDs  []string
-
-		// replication/ transfer / timer task
-		InsertReplicationTasks []Task
-		InsertTransferTasks    []Task
-		InsertTimerTasks       []Task
+		// workflow to be resetted
+		ResetWorkflowSnapshot WorkflowSnapshot
```
Comment (meiliang86) [inline common/persistence/dataInterfaces.go]: `* WorkflowSnapshot`?

File: common/persistence/dataInterfaces.go [hunk 4]
```diff
@@ -849,19 +818,57 @@ type (
...
-		InsertTimerTasks          []Task
-		InsertReplicationTasks    []Task
-		InsertActivityInfos       []*ActivityInfo
-		InsertTimerInfos          []*TimerInfo
-		InsertChildExecutionInfos []*ChildExecutionInfo
-		InsertRequestCancelInfos  []*RequestCancelInfo
-		InsertSignalInfos         []*SignalInfo
-		InsertSignalRequestedIDs  []string
-		Encoding                  common.EncodingType // optional binary encoding type
+		NewWorkflowSnapshot WorkflowSnapshot
```
Comment (meiliang86) [inline common/persistence/dataInterfaces.go]: `* WorkflowSnapshot `?

File: common/persistence/executionStore.go [hunk 5]
```diff
@@ -528,62 +491,66 @@ func (m *executionManagerImpl) ResetMutableState(request *ResetMutableStateReque
...
 				"replication state and version histories cannot both be set."),
 		}
 	}
-	if request.VersionHistories != nil && (request.ExecutionInfo.BranchToken != nil || request.ExecutionInfo.NextEventID != common.EmptyEventID) {
+	if request.ResetWorkflowSnapshot.VersionHistories != nil &&
+		(request.ResetWorkflowSnapshot.ExecutionInfo.BranchToken != nil ||
+			request.ResetWorkflowSnapshot.ExecutionInfo.NextEventID != common.EmptyEventID) {
+
 		return &workflow.InternalServiceError{
 			Message: fmt.Sprintf("Reset workflow with invalid request:" +
```
Comment (meiliang86) [inline common/persistence/executionStore.go]: No need for sprintf? Same in other places

File: service/history/nDCHistoryReplicator.go [hunk 6]
```diff
@@ -256,7 +256,7 @@ func (r *nDCHistoryReplicator) applyStartEvents(
 		return nil
 	}
 
-	currentLastEventID, currentLastWriteVersion, currentLastEventTaskID, err := r.getWorkflowVersionTaskID(
+	currentLastEventID, currentLastWriteVersion, currentLastEventTaskID, err := r.getWorkflowLastEventIDVersionTaskID(
```
Comment (meiliang86) [inline service/history/nDCHistoryReplicator.go]: Might just create a struct

File: service/history/nDCTransactioniMgr.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+		currentContext      workflowExecutionContext
+		currentMutableState mutableState
+
+		metricsClient metrics.Client
+		logger        log.Logger
+	}
+)
+
+func newNDCTransactionMgr(
```
Comment (meiliang86) [inline service/history/nDCTransactioniMgr.go]: Defer this to next PR?

File: common/persistence/dataInterfaces.go (hunk 3 above)
Comment (wxing1292) [reply common/persistence/dataInterfaces.go]: no this one is required.
the *WorkflowSnapshot / *WorkflowMutation is optoinal

File: service/history/nDCHistoryReplicator.go (hunk 6 above)
Comment (wxing1292) [reply service/history/nDCHistoryReplicator.go]: the thing is there are too many struct

File: service/history/nDCTransactioniMgr.go (hunk 7 above)
Comment (wxing1292) [reply service/history/nDCTransactioniMgr.go]: plz review #2115 instead

keeping this refactor in master maybe more preferable

**Comment by wxing1292** [comment]
close ini favor of #2115

**Pull request closed without merging**